src/emulator/hardware/builtin-gates/MipsAlu.js

```diff
@@ -0,0 +1,92 @@
+/**
+ * The MIT License (MIT)
+ * Copyright (c) 2017-present Dmitry Soshnikov <dmitry.soshnikov@gmail.com>
+ */
+
+'use strict';
+
+const BuiltInGate = require('../BuiltInGate');
+
+/**
+ * Canonical truth table for the MipsAlu
+ */
+const TRUTH_TABLE = [
+
+  // Add
+  {a: 1, b: 1, na: 0, nb: 0, less: 0, cin: 0, op: 0b10, out: 0, cout: 1, set: 0 },
+  {a: 1, b: 1, na: 0, nb: 0, less: 0, cin: 1, op: 0b10, out: 1, cout: 1, set: 0 },
+  {a: 0, b: 0, na: 0, nb: 0, less: 0, cin: 1, op: 0b10, out: 1, cout: 0, set: 0 },
+
+  // Sub
+  {a: 1, b: 1, na: 0, nb: 1, less: 0, cin: 1, op: 0b10, out: 0, cout: 1, set: 0 },
+
+  // And
+  {a: 1, b: 1, na: 0, nb: 0, less: 0, cin: 0, op: 0b00, out: 1, cout: 0, set: 0 },
+
+  // Or
+  {a: 1, b: 1, na: 0, nb: 0, less: 0, cin: 0, op: 0b01, out: 1, cout: 0, set: 0 },
+  {a: 0, b: 1, na: 0, nb: 0, less: 0, cin: 0, op: 0b01, out: 1, cout: 0, set: 0 },
+  {a: 0, b: 0, na: 0, nb: 0, less: 0, cin: 0, op: 0b01, out: 0, cout: 0, set: 0 },
+
+];
+
+/**
+ * A 1-bit MipsAlu
+ */
+class MipsAlu extends BuiltInGate {
+  /**
+   * Logic:
+   *
+   *   a & b: op = 0
+   *   a | b: op = 1
+   *   a NOR b => ~a & ~b: op = 0, na = 1, nb = 1
+   *
+   * Math:
+   *
+   *   a + b: op = 2
+   *   a - b: op = 2, nb = 1, cin = 1
+   *   b - a: op = 2, na = 1, cin = 1
+   *
+   */
+  eval() {
+    const [ a, b, na, nb, less, cin, op ] = this.getInputPins().map(p => p.getValue());
+
+    const A = na ? +!a : a;
+    const B = nb ? +!b : b;
+
+    switch(op) {
+      case 0b00: {
+        this.getOutputPins()[0].setValue(A & B); 
+        break;
+      }
+      case 0b01: {
+        this.getOutputPins()[0].setValue(A | B); 
+        break;
+      }
+      case 0b10: {
+        const sum = A + B + cin;
+        this.getOutputPins()[0].setValue(sum % 2); 
+        this.getOutputPins()[1].setValue(Math.floor(sum / 2)); 
```

Comment:
Hm.. in the original [FullAdder](https://github.com/DmitrySoshnikov/hdl-js/blob/master/src/emulator/hardware/builtin-gates/FullAdder.js#L44) I used `Math.trunc` here. Need to double-check which is correct.

Reply:
I think they are equivalent in the 1-bit case here:

```
Math.floor(0 / 2) = 0 = Math.trunc(0 / 2)
Math.floor(1 / 2) = 0 = Math.trunc(1 / 2)
Math.floor(2 / 2) = 1 = Math.trunc(2 / 2)
Math.floor(3 / 2) = 1 = Math.trunc(3 / 2)
```

In general I think they should only differ for negative values, i.e. `Math.floor(-1.5) === -2` but `Math.trunc(-1.5) === -1`.

Reply:
Yeah, technically we do support negative numbers. Should it be `trunc` then? In any case, we need to unify implementation with in both places.